File: java/chaincue-real-estate/src/main/java/com/chaincuerealestate/chaincuerealestate/services/DTOBuilderHelpers/CountryHelper.java
package com.chaincuerealestate.chaincuerealestate.services.DTOBuilderHelpers;

import com.chaincuerealestate.chaincuerealestate.domains.Country;
import com.chaincuerealestate.chaincuerealestate.domains.House;
import com.chaincuerealestate.chaincuerealestate.services.CountryService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.function.BiConsumer;
import java.util.function.Function;

@RequiredArgsConstructor
@Component
public class CountryHelper {
    private final CountryService countryService;

    public <B> Function<B, B> updateDTOBuilderWithCountries(BiConsumer<B, List<Country>> setCountries) {
        return dtoBuilder -> {
            var countries = countryService.findAll();
            setCountries.accept(dtoBuilder, countries);
            return dtoBuilder;
        };
    }
}

File: kotlin/chaincue-real-estate/src/main/kotlin/com/example/chaincuerealestate/services/DTOBuilderHelpers/CountryHelper.kt
package com.example.chaincuerealestate.services.DTOBuilderHelpers

import com.example.chaincuerealestate.domains.Country
import com.example.chaincuerealestate.services.CountryService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

@Component
class CountryHelper(@Autowired private val countryService: CountryService) {

    fun <B> updateDTOBuilderWithCountries(setCountries: (B, List<Country>) -> Unit): (B) -> B {
        return { dtoBuilder ->
            val countries = countryService.findAll()
            setCountries(dtoBuilder, countries)
            dtoBuilder
        }
    }

}
